### ISSUE TITLE
aws_lambda_function always updates even with no changes

### ISSUE THREAD

**Issue by gtirloni** [hashicorp/terraform-provider-aws#9786]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.12.6
terraform-provider-aws 2.23.0

### Affected Resource(s)

aws_lambda_function

### Terraform Configuration Files

```hcl
resource "aws_lambda_function" "default" {
  filename = substr(
    "${path.module}/${local.lambda_filename}-${var.environment}.zip",
    length(path.cwd) + 1,
    -1,
  ) // +1 for removing the "/"
  function_name = local.name
  role          = aws_iam_role.default.arn
  handler       = "lambda_function.lambda_handler"

  runtime = "python3.7"
  source_code_hash = filebase64sha256(
    "${path.module}/${local.lambda_filename}-${var.environment}.zip",
  )
  timeout = "5"
  publish = "true"
  tags = local.tags

  lifecycle {
    ignore_changes = [
      filename,
      last_modified,
    ]
  }
}
```

### Expected Behavior

Terraform should not try to update the resource.

### Actual Behavior

Terraform always wants to update the resource, regardless if there are changes or not.

### Debug output

```
  # module.xx.aws_lambda_function.default will be updated in-place
  ~ resource "aws_lambda_function" "default" {
        arn                            = "arn:aws:lambda:us-east-1:12345678:function:xx"
        filename                       = ".terraform/modules/xx/xx.zip"
        function_name                  = "xx"
        handler                        = "lambda_function.lambda_handler"
        id                             = "xx"
        invoke_arn                     = "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/arn:aws:lambda:us-east-1:12345678:function:xx/invocations"
        last_modified                  = "2019-08-08T11:31:32.783+0000"
        layers                         = []
        memory_size                    = 128
        publish                        = true
      ~ qualified_arn                  = "arn:aws:lambda:us-east-1:12345678:function:xx:1" -> (known after apply)
        reserved_concurrent_executions = -1
        role                           = "arn:aws:iam::12345678:role/xx"
        runtime                        = "python3.7"
        source_code_hash               = "xxxx"
        source_code_size               = 953
        timeout                        = 10
      ~ version                        = "1" -> (known after apply)

        tracing_config {
            mode = "PassThrough"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
### Steps to Reproduce

1. `terraform apply`

### Important Factoids

A workaround is to ignore `version` and `qualified_arn`.

```hcl
  lifecycle {
    ignore_changes = [
      filename,
      last_modified,
      qualified_arn,
      version,
    ]
  }
```

**Comment by canhnp** [comment]
i have the same issue. how to solve it?

**Comment by danieladams456** [comment]
In the latest TF
```
Terraform v0.14.0-beta2
+ provider registry.terraform.io/hashicorp/aws v3.13.0
```

I now get this error since it is a computed attribute.
```log
Error: Cannot ignore argument not set in the configuration

  on ../../modules/lambda-function/lambda.tf line 56, in resource "aws_lambda_function" "lambda":
  56:       qualified_arn,

The ignore_changes argument is not set in the configuration.
The ignore_changes mechanism only applies to changes within the configuration,
and must be used with arguments set in the configuration and not computed by
the provider.
```

**Comment by Tiago-SD** [comment]
I have the exact some issue but at least the workaround worked for me on the following versions:
Terraform v0.12.29
+ provider.aws v3.15.0

**Comment by willhughes-au** [comment]
It's been over 18 months since this was first reported and it's still not been either triaged or assigned. 

It'd be great if this could have some attention paid to it.

**Comment by fumantsu** [comment]
I can also confirm this with provider.aws 3.21, terraform 0.13 with data.archive_file as source and using the outputsha256 as source in the layer.

Really annoying if you don't to changes since it is changing the version as well.

**Comment by chenjie** [comment]
I can confirm that the same exact issue exists on 3.23.0

**Comment by xmackex** [comment]
Issue exists

```
Terraform v0.14.3
+ provider registry.terraform.io/hashicorp/aws v3.25.0
```

**Comment by nickjmv** [comment]
+1

**Comment by reedflinch** [comment]
Unfortunately the workaround does not work if you use a Lambda alias pointing to your function version. Ignoring changes to `version` means the alias resource will not detect changes.

```hcl
resource "aws_lambda_alias" "main" {
  name             = "LIVE"
  function_name    = aws_lambda_function.main.arn
  function_version = aws_lambda_function.main.version
  description      = var.tag
}
```

**Comment by kitos9112** [comment]
This is probably the root cause or may be intrinsically correlated to !17385

**Comment by deyvsh** [comment]
Optimistic that this might be fixed by https://github.com/hashicorp/terraform-provider-aws/pull/17610

**Comment by bill-rich** [comment]
Hi All,
I haven't been able to reproduce this issue with the provided configuration. Is anyone still experiencing this problem? If so, please post a copy of your configuration to help with reproducing the issue.

**Comment by eschulma** [comment]
I had this problem with 3.31. But updating to 3.57 fixed it.

**Comment by breathingdust** [comment]
:wave: It looks like this may have been fixed by either #17385 or #17610, I will close this issue for now, please comment to reopen if this continues to be an issue. Thanks!
